src/adapter/src/coord.rs

```diff
@@ -1487,6 +1495,77 @@ impl Coordinator {
         Ok(())
     }
 
+    /// Returns an `as_of` suitable for bootstrapping the given index dataflow.
+    fn bootstrap_index_as_of(
+        &self,
+        cluster_id: ComputeInstanceId,
+        dataflow: &DataflowDescription<OptimizedMirRelationExpr>,
+    ) -> Antichain<Timestamp> {
+        // All inputs must be readable at the chosen `as_of`, so it must be at least the join of
+        // the `since`s of all dependencies.
+        let storage_ids = dataflow.source_imports.keys().copied().collect();
+        let compute_ids = dataflow.index_imports.keys().copied().collect();
+        let id_bundle = CollectionIdBundle {
+            storage_ids,
+            compute_ids: btreemap! {cluster_id => compute_ids},
+        };
+        let mut as_of = self.least_valid_read(&id_bundle);
+
+        // For compute reconciliation to recognize that an existing dataflow can be reused, we want
+        // to advance the `as_of` far enough that it is beyond the `as_of`s of all dataflows that
+        // might still be installed on replicas, but ideally not much farther as that would just
+        // increase the wait time until the index becomes readable. We advance the `as_of` to the
+        // meet of the `upper`s of all dependencies, as we know that no replica can have produced
+        // output for that time, so we can assume that no replica `as_of` has been adanced beyond
+        // this time either.
+        let write_frontier = self.least_valid_write(&id_bundle);
+        // Things go wrong if we try to create a dataflow with `as_of = []`, so avoid that.
```

Comment:
Not sure if this is the right place to mention it, but what confused me more than once is which objects have a `[]` write frontier. The answer seems to be constant collections.

Reply:
Constant collections are one example. Another are load generator sources that produce a bunch of data and then shut down (I think TPC-H does this). There might be other cases now or in the future, so it's hard to come up with and exhaustive enumeration that isn't at the risk of becoming outdated.

Reply:
Thanks for the explanation!